src/CompanionPacketDecoder.cs

```diff
@@ -0,0 +1,378 @@
+﻿using System;
+using System.Linq;
+using Google.Protobuf;
+using Microsoft.Extensions.Logging;
+
+namespace ZwiftPacketMonitor
+{
+    public class CompanionPacketDecoder
+    {
+        private readonly ILogger<CompanionPacketDecoder> _logger;
+        private readonly IMessageWriter _invidividualFileMessageWriter;
+
+        public CompanionPacketDecoder(IMessageWriter invidividualFileMessageWriter, ILogger<CompanionPacketDecoder> logger)
+        {
+            _invidividualFileMessageWriter = invidividualFileMessageWriter ?? new NopMessageWriter();
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Raised when the Zwift desktop app indicates that a command has become available to the companion app
+        /// </summary>
+        public event EventHandler<CommandAvailableEventArgs> CommandAvailable;
+
+        /// <summary>
+        /// Raised when the companion app sends a command to the desktop app
+        /// </summary>
+        public event EventHandler<CommandSentEventArgs> CommandSent;
+
+        /// <summary>
+        /// Raised when the position of the rider is sent to the companion app
+        /// </summary>
+        public event EventHandler<RiderPositionEventArgs> RiderPosition;
+
+        /// <summary>
+        /// Raised when a power up is rewarded in the game
+        /// </summary>
+        public event EventHandler<PowerUpEventArgs> PowerUp;
+
+        /// <summary>
+        /// Raised when a heart beat message is received
+        /// </summary>
+        public event EventHandler<EventArgs> HeartBeat;
+
+        /// <summary>
+        /// Raised when activity details are received
+        /// </summary>
+        public event EventHandler<ActivityDetailsEventArgs> ActivityDetails;
+
+        public void DecodeOutgoing(byte[] buffer, uint sequenceNumber)
+        {
+            var message = ZwiftCompanionToApp.Parser.ParseFrom(buffer);
+
+            // This is for messages that have only tag1 and tag10
+            // and we can't figure out based on that alone if there
+            // is any proper data in it...
+            if (buffer.Length <= 10)
+            {
+                OnHeartBeat();
+                return;
+            }
+
+            var riderMessage = ZwiftCompanionToAppRiderMessage.Parser.ParseFrom(buffer);
+
+            if (riderMessage.Details == null && message.Tag10 == 0)
+            {
+                var typeTag10Zero = ZwiftCompanionToAppMessageTag10Zero.Parser.ParseFrom(buffer);
+                var clockTime = DateTimeOffset.FromUnixTimeSeconds((long)typeTag10Zero.ClockTime);
+                _logger.LogDebug("Sent a tag 10 = 0 type message with timestamp {clock_time}", clockTime);
```

Comment:
I’ve gone the structured logging approach here actually. Advantage of that is that when you log to JSON for example instead of a console you get access to the placeholder which makes log analytics simpler. 

If you prefer I can change it to string interpolation.

Reply:
I've brought this into line with the existing calls to LogInfo/LogDebug etc.